hikari/internal/data_binding.py

```diff
@@ -130,36 +130,58 @@ def __init__(
         super().__init__(dumps(value), content_type=_JSON_CONTENT_TYPE, encoding=_UTF_8)
 
 
+class _FilePayload(aiohttp.Payload):
+    _value: files.Resource[files.AsyncReader]
+
+    def __init__(
+        self,
+        value: files.Resource[files.AsyncReader],
+        content_type: str,
+        /,
+        *,
+        executor: typing.Optional[concurrent.futures.Executor] = None,
+        headers: typing.Optional[typing.Dict[str, str]] = None,
+    ) -> None:
+        super().__init__(value=value, headers=headers, content_type=content_type)
+        self._executor = executor
+
+    async def write(self, writer: aiohttp.abc.AbstractStreamWriter) -> None:
+        async with self._value.stream(executor=self._executor) as data:
+            async for chunk in data:
+                await writer.write(chunk)
+
+
 @typing.final
 class URLEncodedFormBuilder:
     """Helper class to generate `aiohttp.FormData`."""
 
     __slots__: typing.Sequence[str] = ("_fields", "_resources")
 
     def __init__(self) -> None:
-        self._fields: typing.List[typing.Tuple[str, typing.Union[str, bytes], typing.Optional[str]]] = []
+        self._fields: typing.List[typing.Tuple[str, bytes, typing.Optional[str]]] = []
         self._resources: typing.List[typing.Tuple[str, files.Resource[files.AsyncReader]]] = []
 
-    def add_field(
-        self, name: str, data: typing.Union[str, bytes], *, content_type: typing.Optional[str] = None
-    ) -> None:
+    def add_field(self, name: str, data: bytes, *, content_type: typing.Optional[str] = None) -> None:
         self._fields.append((name, data, content_type))
 
     def add_resource(self, name: str, resource: files.Resource[files.AsyncReader]) -> None:
         self._resources.append((name, resource))
 
-    async def build(
-        self, stack: contextlib.AsyncExitStack, executor: typing.Optional[concurrent.futures.Executor] = None
-    ) -> aiohttp.FormData:
-        form = aiohttp.FormData()
+    async def build(self, executor: typing.Optional[concurrent.futures.Executor] = None) -> aiohttp.MultipartWriter:
+        form = aiohttp.MultipartWriter(subtype="form-data")
 
         for field in self._fields:
-            form.add_field(field[0], field[1], content_type=field[2])
+            body_payload = aiohttp.BytesPayload(field[1], content_type=field[2])
+            body_payload.set_content_disposition("form-data", name=field[0])
+            form.append_payload(body_payload)
 
         for name, resource in self._resources:
-            stream = await stack.enter_async_context(resource.stream(executor=executor))
-            mimetype = stream.mimetype or _APPLICATION_OCTET_STREAM
-            form.add_field(name, stream, filename=stream.filename, content_type=mimetype)
+            async with resource.stream(executor=executor, head_only=True) as stream:
```

Comment:
maybe this could be optimised a tiny bit by adding a get_mimetype method to Resource which is specialised for impls where we already known the mimetype before reading to not do a head_only stream or something